site/utils/apiUtils/index.ts

```diff
@@ -107,3 +115,205 @@ export const getReturnPage = (req: NextApiRequest) => {
+
+export const publishToHootsuite = async (socialMediaPosts: SocialMediaPost[], orgId: string) => {
+    await Promise.all(
+        socialMediaPosts
+            .filter((s) => s.status === SocialMediaPostStatus.pending)
+            .map(async (socialMediaPost) => {
+                const refreshTokens = await getParametersByPath(`/social/${orgId}/hootsuite`);
+
+                if (!refreshTokens || (refreshTokens && refreshTokens.Parameters?.length === 0)) {
+                    await upsertSocialMediaPost(
+                        {
+                            ...socialMediaPost,
+                            status: SocialMediaPostStatus.rejected,
+                        },
+                        orgId,
+                    );
+                    logger.debug("Refresh token is required when creating a social media post");
+                }
+                const refreshToken = refreshTokens.Parameters?.find((rt) =>
+                    rt.Name?.includes(`${socialMediaPost.socialAccount}`),
+                );
+                if (!refreshToken) {
+                    await upsertSocialMediaPost(
+                        {
+                            ...socialMediaPost,
+                            status: SocialMediaPostStatus.rejected,
+                        },
+                        orgId,
+                    );
+                    logger.debug("Refresh token is required when creating a social media post");
+                }
+                const [clientId, clientSecret] = await Promise.all([
+                    getParameter(`/social/hootsuite/client_id`),
+                    getParameter(`/social/hootsuite/client_secret`),
+                ]);
+
+                if (!clientId || !clientSecret) {
+                    await upsertSocialMediaPost(
+                        {
+                            ...socialMediaPost,
+                            status: SocialMediaPostStatus.rejected,
+                        },
+                        orgId,
+                    );
+                    logger.debug("clientId and clientSecret must be defined");
+                }
+
+                const credentials = `${clientId.Parameter?.Value || ""}:${clientSecret.Parameter?.Value || ""}`;
+
+                const authToken = `Basic ${Buffer.from(credentials).toString("base64")}`;
+                const responseToken = await fetch(`${HOOTSUITE_URL}oauth2/token`, {
+                    method: "POST",
+                    body: new URLSearchParams({
+                        grant_type: "refresh_token",
+                        refresh_token: refreshToken?.Value ?? "",
+                    }),
+                    headers: {
+                        "Content-Type": "application/x-www-form-urlencoded",
+                        Authorization: authToken,
+                    },
+                });
+
+                if (responseToken.ok) {
+                    const tokenResult = hootsuiteTokenSchema.parse(await responseToken.json());
```

Comment:
Could this value be obtained once? Currently hootsuite api's are called for every post in pending. We need the access token only once in this iteration of pending social media posts as the access token will be valid for some time.

Reply:
it could be made with a different account so hence it needs to be called for every post